File: Assignment/3학년 2학기/Python/중간고사/Client/Client/ServerPackage/Packing.py
from ServerPackage import Socket
from Global import *
import struct

class CPacking(Socket.CSocket):
    # 상속 받은 객체 초기화, 프로토콜, 수신 데이터 변수
    def __init__(self):
        Socket.CSocket.__init__(self)
        self.protocol = 0
        self.recvied = 0

    # 전체 사이즈 언패킹
    def size_unpacking(self, buffer):
        unpack = struct.unpack('=i', buffer)
        return unpack[0]

    # 프로토콜 언패킹
    def protocol_unpacking(self, buffer):
        unpack = struct.unpack('=i', buffer)
        return unpack[0]

    #def unpacking(self, buffer):
    #    struct_fmt = str.format('={0}s', (len(buffer)))
    #    unpack = struct.unpack(struct_fmt, buffer)
    #    return unpack   

    # 송신
    def Complete_Send(self, data):
        self.Send(data)       

    # 수신
    def Complete_Recv(self):
        # Size 받기
        self.recvied = self.Recv(4)
        if not self.recvied:
            return False        
        size = self.size_unpacking(self.recvied)

        # Protocol 받기
        self.recvied = self.Recv(4)
        if not self.recvied:
            return False

        self.protocol = self.protocol_unpacking(self.recvied)
        size = size - 4

        # Protocol 만 받은 경우 데이터가 없기 때문에 예외 처리
        if size == 0:
            return True

        # 데이터 받기
        self.recvied = self.Recv(size)
        if not self.recvied:
            return False

        return True

File: Assignment/3학년 1학기/NetworkPrograming (Deepening)/IOCP_Class/Packing.cpp
#include "Packing.h"
#include"Crypt.h"

CPacking::CPacking()
{
	recvbytes = 0;
	comp_recvbytes = 0;
	ZeroMemory(recvbuf, sizeof(recvbuf));

	recv_sizeflag = false;

	sendbuf = new queue<QueueSendData*>();
}

CPacking::CPacking(SOCKET _sock) : CTCPSocket(_sock)
{
	recvbytes = 0;
	comp_recvbytes = 0;
	ZeroMemory(recvbuf, sizeof(recvbuf));

	recv_sizeflag = false;

	sendbuf = new queue<QueueSendData*>();
}

CPacking::~CPacking()
{
	delete sendbuf;
}

void CPacking::PackingData(PROTOCOL protocol, const char* buf, int _size)
{
	CLock lock;
	char encrypt[4096];
	ZeroMemory(encrypt, sizeof(encrypt));
	QueueSendData* data = new QueueSendData;
	ZeroMemory(data, sizeof(QueueSendData));
	char* ptr = data->sendbuf;
	int size = 0;

	ptr = ptr + sizeof(size);



	memcpy(ptr, &protocol, sizeof(PROTOCOL));
	ptr = ptr + sizeof(PROTOCOL);
	size = size + sizeof(PROTOCOL);

	if (buf != nullptr)
	{
		memcpy(ptr, buf, _size);
		size = size + _size;
	}

	CCrypt::Decrypt((BYTE*)buf, (BYTE*)encrypt, size);
	//데이터 첫번째 전체 패킷 길이를 제외한 packing이 다된 상태에서 암호화를 진행 해야 하므로.

	ptr = data->sendbuf;
	memcpy(ptr, &size, sizeof(int));

	size = size + sizeof(int);
	data->sendbytes = size;

	sendbuf->push(data);
}

bool CPacking::Send()
{
	CLock lock;

	if (sendbuf->size() > 1)
	{
		return true;
	}

	return WSA_Send(sendbuf->front()->sendbuf + sendbuf->front()->comp_sendbytes, sendbuf->front()->sendbytes - sendbuf->front()->comp_sendbytes);
}

int CPacking::CompleteSend(int _completebyte)
{
	CLock lock;

	sendbuf->front()->comp_sendbytes += _completebyte;

	if (sendbuf->front()->comp_sendbytes == sendbuf->front()->sendbytes)
	{
		sendbuf->pop();

		if (sendbuf->empty())
		{
			return SOC_TRUE;
		}
		else
		{
			if (!Send()) return SOC_ERROR;
			return SOC_FALSE;
		}
	}

	if (!Send()) return SOC_ERROR;

	return SOC_FALSE;
}

PROTOCOL CPacking::GetProtocol()
{
	CLock lock;

	PROTOCOL protocol;
	memcpy(&protocol, recvbuf, sizeof(PROTOCOL));
	return protocol;
}

const char* CPacking::UnPackData()
{
	CLock lock;

	const char* ptr = recvbuf + sizeof(PROTOCOL);

	return ptr;
}

bool CPacking::Recv()
{
	CLock lock;

	int len;
	if (recv_sizeflag)
	{
		len = recvbytes - comp_recvbytes;
	}
	else
	{
		len = sizeof(int) - comp_recvbytes;
	}

	return WSA_Recv(recvbuf + comp_recvbytes, len);
}

int CPacking::CompleteRecv(int _completebyte)
{
	CLock lock;
	char decrypt[4096];
	ZeroMemory(decrypt, sizeof(decrypt));
	if (!recv_sizeflag)
	{
		comp_recvbytes += _completebyte;

		if (comp_recvbytes == sizeof(int))
		{
			memcpy(&recvbytes, recvbuf, sizeof(int));
			comp_recvbytes = 0;
			recv_sizeflag = true;
		}

		if (!Recv()) return SOC_ERROR;

		return SOC_FALSE;
	}

	comp_recvbytes += _completebyte;

	if (comp_recvbytes == recvbytes)
	{
		CCrypt::Decrypt((BYTE*)recvbuf, (BYTE*)decrypt, recvbytes);//데이터가 다 받아진 상태에서 복호화를 진행 해야 하므로.
		comp_recvbytes = 0;
		recvbytes = 0;
		recv_sizeflag = false;

		return SOC_TRUE;
	}
	else
	{
		if (!Recv()) return SOC_ERROR;

		return SOC_FALSE;
	}
}